Fix store based barrier to only use 'add'.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49930 Fix store based barrier to only use 'add'.**

Certain store implementations don't work well when we use get() and
add() on the same key. To avoid this issue, we only use add() in the store
based barrier. The buggy store implementations can't be properly fixed due to
legacy reasons.

Differential Revision: [D25725386](https://our.internmc.facebook.com/intern/diff/D25725386/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25725386/)!

pritamdamania committed:
Fix store based barrier to only use 'add'.

Certain store implementations don't work well when we use get() and
add() on the same key. To avoid this issue, we only use add() in the store
based barrier. The buggy store implementations can't be properly fixed due to
legacy reasons.

Differential Revision: [D25725386](https://our.internmc.facebook.com/intern/diff/D25725386/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25725386/)!

[ghstack-poisoned]

pritamdamania committed:
Update on "Fix store based barrier to only use 'add'."


Certain store implementations don't work well when we use get() and
add() on the same key. To avoid this issue, we only use add() in the store
based barrier. The buggy store implementations can't be properly fixed due to
legacy reasons.

Differential Revision: [D25725386](https://our.internmc.facebook.com/intern/diff/D25725386/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25725386/)!

[ghstack-poisoned]

pritamdamania committed:
Update on "Fix store based barrier to only use 'add'."


Certain store implementations don't work well when we use get() and
add() on the same key. To avoid this issue, we only use add() in the store
based barrier. The buggy store implementations can't be properly fixed due to
legacy reasons.

Differential Revision: [D25725386](https://our.internmc.facebook.com/intern/diff/D25725386/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25725386/)!

[ghstack-poisoned]

osalpekar commented on torch/distributed/distributed_c10d.py:
```diff
@@ -191,16 +183,25 @@ def _store_based_barrier(rank, store, timeout):
+    # doesn't work well with 'get'. Ideally the store implementations should
+    # be fixed, but for backward compatiblity reasons it is risky to change
+    # the store implementations. Once, we completely migrate away from these
+    # legacy stores, we can use 'get' here instead.
```
Curious - which store implementations display this behavior, and what happens when 'get' is called? Since the store docs were only added in 1.7, I'm guessing not that many OSS users use the store directly so we should be alright fixing the store implementations wrt backward compatibility right?

osalpekar replied:
Never mind, just saw your comment regarding the Caffe2 stores

osalpekar reviewed (approved):
Thanks for the fix!

rohan-varma commented on torch/distributed/distributed_c10d.py:
```diff
@@ -191,16 +183,25 @@ def _store_based_barrier(rank, store, timeout):
     while worker_count != world_size:
         time.sleep(0.01)
-        worker_count = int(store.get(store_key))
+        worker_count = store.add(store_key, 0)
```
I'm assuming the timeout is reasonably large when we see "timed out initializing process group", can we add some logging here showing the number of workers (`worker_count`) every 10-30 seconds? Would probably also help with the debugging of the issue we're seeing.

rohan-varma commented on torch/distributed/distributed_c10d.py:
```diff
@@ -191,16 +183,25 @@ def _store_based_barrier(rank, store, timeout):
         if timedelta(seconds=(time.time() - start)) > timeout:
-            raise RuntimeError("Timed out initializing process group")
+            raise RuntimeError(
+                "Timed out initializing process group in store based barrier on "
```
Can we log the timeout here if it's easily available?

rohan-varma commented on torch/distributed/distributed_c10d.py:
```diff
@@ -191,16 +183,25 @@ def _store_based_barrier(rank, store, timeout):
     while worker_count != world_size:
         time.sleep(0.01)
-        worker_count = int(store.get(store_key))
+        worker_count = store.add(store_key, 0)
```
Also just noticed, previously this line had `int` cast but now with `add` it does not, is this a bug or do `get` and `add` return different types (str vs int) and is that intentional design of the store api (cc @osalpekar)?

osalpekar replied:
This is actually the expected behavior - the default type for the value when `set`/`get` are used is `const std::vector<uint8_t>&` so we must cast to int. In the `add` case, we're simply bumping a counter so the underlying value is stored as an `int64_t` (which is what add returns).

pritamdamania committed:
Update on "Fix store based barrier to only use 'add'."


Certain store implementations don't work well when we use get() and
add() on the same key. To avoid this issue, we only use add() in the store
based barrier. The buggy store implementations can't be properly fixed due to
legacy reasons.

Differential Revision: [D25725386](https://our.internmc.facebook.com/intern/diff/D25725386/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25725386/)!

[ghstack-poisoned]
